File: src/main/java/com/prog5/items/Wallet.java
package com.prog5.items;

import com.prog5.items.utils.WalletUtils;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Scanner;

public class Wallet {
    private String owner;

    private boolean isLost;

    private List<IDCard> idCards;

    private List<Money> moneyList;

    private List<CreditCard> creditCards;

    private List<BusinessCard> businessCards;

    private List<DrivingLicence> drivingLicences;

    private List<IDPhoto> idPhotos;

    public Wallet(
            String owner,
            boolean isLost,
            List<IDCard> idCards,
            List<Money> moneyList,
            List<CreditCard> creditCards,
            List<DrivingLicence> drivingLicences,
            List<IDPhoto> idPhotos,
            List<BusinessCard> businessCards) {
        this.owner = owner;
        this.isLost = isLost;
        this.idCards = idCards;
        this.moneyList = moneyList;
        this.creditCards = creditCards;
        this.drivingLicences = drivingLicences;
        this.idPhotos = idPhotos;
        this.businessCards = businessCards;
    }

    public String getOwner() {
        return owner;
    }

    public boolean isLost() {
        return isLost;
    }

    public void setLost() {
        isLost = !isLost;
    }

    public List<IDCard> getIdCards() {
        return idCards;
    }

    Scanner scanner = new Scanner(System.in);

    public void addIdCard() {
        if (!isLost) {
            System.out.println("Enter ID Card name: ");
            String idCardName = scanner.nextLine();

            System.out.print("Enter ID Card number: ");
            String idCardNumber = scanner.nextLine();

            IDCard newIdCard = new IDCard(idCardName, idCardNumber, false);

            idCards.add(newIdCard);

            System.out.println("ID Card added to the wallet.");
        } else {
            System.out.println("Cannot add ID Card. The wallet is lost.");
        }
    }

    public List<Money> getMoneyList() {
        return moneyList;
    }

    public void addMoney() {
        if (!isLost) {
            System.out.print("Enter the amount of money: ");
            double amount = scanner.nextDouble();
            scanner.nextLine();

            System.out.println("Choose currency:");
            System.out.println("\n1. Dollars\n2. Euro\n3. Ariary");
            System.out.print("\nEnter currency choice (1-3): ");
            int currencyChoice = scanner.nextInt();
            scanner.nextLine();

            Money.Currency currency;
            switch (currencyChoice) {
                case 1 -> currency = Money.Currency.Dollars;
                case 2 -> currency = Money.Currency.Euro;
                case 3 -> currency = Money.Currency.Ariary;
                default -> {
                    System.out.println("Invalid currency choice. Using Dollars by default.");
                    currency = Money.Currency.Dollars;
                }
            }

            Money money = new Money(currency, amount);
            moneyList.add(money);
            System.out.println("Money added to the wallet.");
        } else {
            System.out.println("Cannot add money. The wallet is lost.");
        }
    }

    public List<CreditCard> getCreditCards() {
        return creditCards;
    }

    public void addCreditCard() {
        if (!isLost) {
            System.out.print("Enter Credit Card owner: ");
            String cardOwner = scanner.nextLine();

            System.out.print("Enter Credit Card number: ");
            String cardNumber = scanner.nextLine();

            System.out.print("Enter expiration date (yyyy-MM-dd): ");
            String dateString = scanner.nextLine();
            LocalDate expirationDate = LocalDate.parse(dateString, DateTimeFormatter.ISO_DATE);

            CreditCard newCreditCard = new CreditCard(cardOwner, cardNumber, expirationDate);
            creditCards.add(newCreditCard);

            System.out.println("Credit Card added to the wallet.");
        } else {
            System.out.println("Cannot add Credit Card. The wallet is lost.");
        }
    }

    public List<DrivingLicence> getDrivingLicences() {
        return drivingLicences;
    }

    public void addDrivingLicence() {
        if (!isLost) {
            System.out.print("Enter Driving Licence owner: ");
            String licenceOwner = scanner.nextLine();

            System.out.println("Choose Driving Licence category:");
            System.out.println("1. A");
            System.out.println("2. B");
            System.out.println("3. C");
            System.out.println("4. D");
            System.out.println("5. E");
            System.out.print("Enter category choice (1-5): ");
            int categoryChoice = scanner.nextInt();
            scanner.nextLine();

            DrivingLicence.Category licenceCategory;
            switch (categoryChoice) {
                case 1 -> licenceCategory = DrivingLicence.Category.A;
                case 2 -> licenceCategory = DrivingLicence.Category.B;
                case 3 -> licenceCategory = DrivingLicence.Category.C;
                case 4 -> licenceCategory = DrivingLicence.Category.D;
                case 5 -> licenceCategory = DrivingLicence.Category.E;
                default -> {
                    System.out.println("Invalid category choice. Using A by default.");
                    licenceCategory = DrivingLicence.Category.A;
                }
            }

            DrivingLicence newDrivingLicence = new DrivingLicence(licenceOwner, licenceCategory);
            drivingLicences.add(newDrivingLicence);

            System.out.println("Driving Licence added to the wallet.");
        } else {
            System.out.println("Cannot add Driving Licence. The wallet is lost.");
        }
    }

    public List<IDPhoto> getIdPhotos() {
        return idPhotos;
    }

    public void addIdPhoto() {
        if (!isLost) {
            System.out.print("Enter ID Photo owner: ");
            String photoOwner = scanner.nextLine();

            IDPhoto newIdPhoto = new IDPhoto(photoOwner);
            idPhotos.add(newIdPhoto);

            System.out.println("ID Photo added to the wallet.");
        } else {
            System.out.println("Cannot add ID Photo. The wallet is lost.");
        }
    }

    public List<BusinessCard> getBusinessCards() {
        return businessCards;
    }

    public void addBusinessCard() {
        if (!isLost) {
            System.out.print("Enter Business Card owner: ");
            String cardOwner = scanner.nextLine();

            System.out.print("Enter Company Name: ");
            String companyName = scanner.nextLine();

            BusinessCard newBusinessCard = new BusinessCard(cardOwner, companyName);
            businessCards.add(newBusinessCard);

            System.out.println("Business Card added to the wallet.");
        } else {
            System.out.println("Cannot add Business Card. The wallet is lost.");
        }
    }

    public void removeItem(WalletUtils.ItemType itemType) {
        switch (itemType) {
            case ID_CARD -> removeItem(getIdCards(), "ID Card");
            case MONEY -> removeItem(getMoneyList(), "Money item");
            case CREDIT_CARD -> removeItem(getCreditCards(), "Credit Card");
            case BUSINESS_CARD -> removeItem(getBusinessCards(), "Business Card");
            case DRIVING_LICENCE -> removeItem(getDrivingLicences(), "Driving Licence");
            case ID_PHOTO -> removeItem(getIdPhotos(), "ID Photo");
        }
    }

    private <T> void removeItem(List<T> itemList, String itemType) {
        if (itemList.isEmpty()) {
            System.out.println("\nNo " + itemType + "s found in the wallet.");
            return;
        }

        System.out.println("\n"+ itemType + "s in the wallet:");
        for (int i = 0; i < itemList.size(); i++) {
            System.out.println("\n" +(i + 1) + ". " + itemList.get(i));
        }

        Scanner scanner = new Scanner(System.in);

        System.out.print("\nEnter the number of the " + itemType + " to remove: ");
        int choice = scanner.nextInt();

        if (choice >= 1 && choice <= itemList.size()) {
            T removedItem = itemList.remove(choice - 1);
            System.out.println("\nRemoved " + itemType + ": " + removedItem);
        } else {
            System.out.println("\nInvalid choice. No " + itemType + " removed.");
        }
    }
}

File: Wallet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;

namespace Items
{
    public class Wallet
    {
        public string Owner { get; set; }
        public bool IsLostStatus { get; set; }
        public List<IDCard> IDCards { get; set; }
        public List<Money> MoneyList { get; set; }
        public List<CreditCard> CreditCards { get; set; }
        public List<DrivingLicence> DrivingLicences { get; set; }
        public List<IDPhoto> IDPhotos { get; set; }
        public List<BusinessCard> BusinessCards { get; set; }

        public Wallet(string owner, bool isLost, List<IDCard> idCards, List<Money> moneyList, List<CreditCard> creditCards, List<DrivingLicence> drivingLicences, List<IDPhoto> idPhotos, List<BusinessCard> businessCards)
        {
            Owner = owner;
            IsLostStatus = isLost;
            IDCards = idCards;
            MoneyList = moneyList;
            CreditCards = creditCards;
            DrivingLicences = drivingLicences;
            IDPhotos = idPhotos;
            BusinessCards = businessCards;
        }

        public string GetOwner()
        {
            return Owner;
        }

        public void SetLost()
        {
            IsLostStatus = !IsLostStatus;
        }

        public List<IDCard> GetIDCards()
        {
            return IDCards;
        }

        public void AddIDCard()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter ID Card name: ");
                string idCardName = Console.ReadLine();
                Console.Write("Enter ID Card number: ");
                string idCardNumber = Console.ReadLine();
                IDCard newIDCard = new IDCard(idCardName, idCardNumber, false);
                IDCards.Add(newIDCard);
                Console.WriteLine("ID Card added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add ID Card. The wallet is lost.");
            }
        }

        public List<Money> GetMoneyList()
        {
            return MoneyList;
        }

        public void AddMoney()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter the amount of money: ");
                double amount = Convert.ToDouble(Console.ReadLine());
                Console.WriteLine("Choose currency:\n1. Dollars\n2. Euro\n3. Ariary");
                int currencyChoice = Convert.ToInt32(Console.ReadLine());
                string[] currencies = { "Dollars", "Euro", "Ariary" };
                string currency = (1 <= currencyChoice && currencyChoice <= 3) ? currencies[currencyChoice - 1] : "Dollars";
                Money money = new Money(currency, Convert.ToDecimal(amount));
                MoneyList.Add(money);
                Console.WriteLine("Money added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add money. The wallet is lost.");
            }
        }

        public List<CreditCard> GetCreditCards()
        {
            return CreditCards;
        }

        public void AddCreditCard()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter Credit Card owner: ");
                string cardOwner = Console.ReadLine();
                Console.Write("Enter Credit Card number: ");
                string cardNumber = Console.ReadLine();
                Console.Write("Enter expiration date (yyyy-MM-dd): ");
                string expirationDateString = Console.ReadLine();
                CreditCard newCreditCard = new CreditCard(cardOwner, cardNumber, expirationDateString);
                CreditCards.Add(newCreditCard);
                Console.WriteLine("Credit Card added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add Credit Card. The wallet is lost.");
            }
        }

        public List<DrivingLicence> GetDrivingLicences()
        {
            return DrivingLicences;
        }

        public void AddDrivingLicence()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter Driving Licence owner: ");
                string licenceOwner = Console.ReadLine();
                Console.WriteLine("Choose Driving Licence category:\n1. A\n2. B\n3. C\n4. D\n5. E");
                int categoryChoice = Convert.ToInt32(Console.ReadLine());
                string[] categories = { "A", "B", "C", "D", "E" };
                string licenceCategory = (1 <= categoryChoice && categoryChoice <= 5) ? categories[categoryChoice - 1] : "A";
                DrivingLicence newDrivingLicence = new DrivingLicence(licenceOwner, licenceCategory);
                DrivingLicences.Add(newDrivingLicence);
                Console.WriteLine("Driving Licence added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add Driving Licence. The wallet is lost.");
            }
        }

        public List<IDPhoto> GetIDPhotos()
        {
            return IDPhotos;
        }

        public void AddIDPhoto()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter ID Photo owner: ");
                string photoOwner = Console.ReadLine();
                IDPhoto newIDPhoto = new IDPhoto(photoOwner);
                IDPhotos.Add(newIDPhoto);
                Console.WriteLine("ID Photo added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add ID Photo. The wallet is lost.");
            }
        }

        public List<BusinessCard> GetBusinessCards()
        {
            return BusinessCards;
        }

        public void AddBusinessCard()
        {
            if (!IsLostStatus)
            {
                Console.Write("Enter Business Card owner: ");
                string cardOwner = Console.ReadLine();
                Console.Write("Enter Company Name: ");
                string companyName = Console.ReadLine();
                BusinessCard newBusinessCard = new BusinessCard(cardOwner, companyName);
                BusinessCards.Add(newBusinessCard);
                Console.WriteLine("Business Card added to the wallet.");
            }
            else
            {
                Console.WriteLine("Cannot add Business Card. The wallet is lost.");
            }
        }

        public void RemoveItemByType(string itemType)
        {
            switch (itemType)
            {
                case "ID_CARD":
                    RemoveItem(GetIDCards(), "ID Card");
                    break;
                case "MONEY":
                    RemoveItem(GetMoneyList(), "Money item");
                    break;
                case "CREDIT_CARD":
                    RemoveItem(GetCreditCards(), "Credit Card");
                    break;
                case "BUSINESS_CARD":
                    RemoveItem(GetBusinessCards(), "Business Card");
                    break;
                case "DRIVING_LICENCE":
                    RemoveItem(GetDrivingLicences(), "Driving Licence");
                    break;
                case "ID_PHOTO":
                    RemoveItem(GetIDPhotos(), "ID Photo");
                    break;
                default:
                    Console.WriteLine("Invalid item type.");
                    break;
            }
        }

        private void RemoveItem<T>(List<T> itemList, string itemType)
        {
            if (!itemList.Any())
            {
                Console.WriteLine($"\nNo {itemType}s found in the wallet.");
                return;
            }

            Console.WriteLine($"\n{itemType}s in the wallet:");
            for (int i = 0; i < itemList.Count; i++)
            {
                Console.WriteLine($"\n{i + 1}. {JsonConvert.SerializeObject(itemList[i])}");
            }

            Console.Write($"\nEnter the number of the {itemType} to remove: ");
            int choice = Convert.ToInt32(Console.ReadLine());

            if (1 <= choice && choice <= itemList.Count)
            {
                T removedItem = itemList[choice - 1];
                itemList.RemoveAt(choice - 1);
                Console.WriteLine($"\nRemoved {itemType}: {JsonConvert.SerializeObject(removedItem)}");
            }
            else
            {
                Console.WriteLine($"\nInvalid choice. No {itemType} removed.");
            }
        }
    }
}
